src/gui/newwizard/SetupWizardController.cpp

```diff
@@ -0,0 +1,266 @@
+    QScopedPointer<AbstractSetupWizardPage> page(_currentPage);
+
+    // initial state
+    if (!currentPage.has_value()) {
+        desiredPage = 0;
+    }
+
+    // "next button" workflow
+    if (!desiredPage.has_value()) {
+        // try to fill in data appropriately
+        // if it works, go to next page
+        // otherwise, show current page again
+        if (currentPage == 0) {
+            const auto *pagePtr = qobject_cast<ServerUrlSetupWizardPage *>(_currentPage);
+
+            auto serverUrl = pagePtr->serverUrl();
+
+            // fix scheme if necessary
+            // the second half is needed for URLs which contain a port (e.g., host:1234), QUrl then parses host: as scheme
+            if (serverUrl.scheme().isEmpty()) {
+                serverUrl = QUrl(QStringLiteral("https://") + serverUrl.toString());
+            }
+
+            // TODO: move validation somewhere else, e.g., the pages
+            // TODO: perform some better validation
+            if (serverUrl.isValid()) {
+                // FIXME: use actual workflow type
+                _accountBuilder.setServerUrl(serverUrl, DetermineAuthTypeJob::AuthType::Basic);
+                desiredPage = currentPage.value() + 1;
+            } else {
+                _wizardWindow->showErrorMessage(QStringLiteral("Invalid server URL"));
+                desiredPage = currentPage.value();
+            }
+        }
+
+        if (currentPage == 1) {
+            if (_accountBuilder.authType() == DetermineAuthTypeJob::AuthType::Basic) {
+                const auto *pagePtr = qobject_cast<BasicCredentialsSetupWizardPage *>(_currentPage);
+
+                const auto username = pagePtr->username();
+                const auto password = pagePtr->password();
+
+                _accountBuilder.setAuthenticationStrategy(new HttpBasicAuthenticationStrategy(username, password));
+
+                // TODO: actually check whether creds are correct
+                if (_accountBuilder.hasValidCredentials()) {
+                    desiredPage = currentPage.value() + 1;
+                } else {
+                    _wizardWindow->showErrorMessage(QStringLiteral("Invalid credentials"));
+                    desiredPage = currentPage.value();
+                }
+            }
+
+            if (_accountBuilder.authType() == DetermineAuthTypeJob::AuthType::OAuth) {
+                // authentication data is filled in asynchronously, hence all we have to do here is determine the next page
+                if (_accountBuilder.hasValidCredentials()) {
+                    desiredPage = currentPage.value() + 1;
+                } else {
+                    _wizardWindow->showErrorMessage(QStringLiteral("Invalid credentials"));
+                    desiredPage = currentPage.value();
+                }
+            }
+        }
+
+        // final step
+        if (currentPage == 2) {
+            const auto *pagePtr = qobject_cast<AccountConfiguredWizardPage *>(_currentPage);
```

Comment:
Could be put in to a helper together with a Q_ASSERT(pagePtr)

Reply:
The cast is not needed anyway, since the controller doesn't need to parse data off the final page.